presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java

```diff
@@ -0,0 +1,307 @@
+import static com.facebook.presto.sql.relational.SqlToRowExpressionTranslator.translate;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Integer.min;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class ExpressionEquivalence
+{
+    private static final Ordering<RowExpression> ROW_EXPRESSION_ORDERING = Ordering.from(new RowExpressionComparator());
+    private static final CanonicalizationVisitor CANONICALIZATION_VISITOR = new CanonicalizationVisitor();
+    private final Metadata metadata;
+    private final SqlParser sqlParser;
+
+    public ExpressionEquivalence(Metadata metadata, SqlParser sqlParser)
+    {
+        this.metadata = metadata;
+        this.sqlParser = sqlParser;
+    }
+
+    public boolean areExpressionsEquivalent(Session session, Expression leftExpression, Expression rightExpression, Map<Symbol, Type> types)
+    {
+        Map<Symbol, Integer> symbolInput = new HashMap<>();
+        Map<Integer, Type> inputTypes = new HashMap<>();
+        int inputId = 0;
+        for (Entry<Symbol, Type> entry : types.entrySet()) {
+            symbolInput.put(entry.getKey(), inputId);
+            inputTypes.put(inputId, entry.getValue());
+            inputId++;
+        }
+        RowExpression leftRowExpression = toRowExpression(session, leftExpression, symbolInput, inputTypes);
+        RowExpression rightRowExpression = toRowExpression(session, rightExpression, symbolInput, inputTypes);
+
+        RowExpression canonicalizedLeft = leftRowExpression.accept(CANONICALIZATION_VISITOR, null);
+        RowExpression canonicalizedRight = rightRowExpression.accept(new CanonicalizationVisitor(), null);
+
+        return canonicalizedLeft.equals(canonicalizedRight);
+    }
+
+    private RowExpression toRowExpression(Session session, Expression expression, Map<Symbol, Integer> symbolInput, Map<Integer, Type> inputTypes)
+    {
+        // check for missing coercions
+        ExpressionAnalysis expressionAnalysis = analyzeExpressionsWithSymbols(
+                session,
+                metadata,
+                sqlParser,
+                symbolInput.keySet().stream()
+                        .collect(Collectors.toMap(identity(), symbol -> inputTypes.get(symbolInput.get(symbol)))),
+                ImmutableList.of(expression));
+
+        // add coercions
+        Expression coercedExpression = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                Expression rewrittenExpression = treeRewriter.defaultRewrite(node, context);
+
+                // cast expression if coercion is registered
+                Type coerceToType = expressionAnalysis.getCoercion(node);
```

Comment:
I thought we already inserted all necessary explicit casts into the Expressions?